### PR TITLE
Handling of server addresses in Console

### PR DESCRIPTION
**Pull request by htdvisser**
#### Summary

This is a quickfix for the console that changes the prefilled gateway server address of the gateway create view to only fill the hostname and omit the port. A similar change was made in the SDK.

#### Changes

- Changed GS/NS/AS/JS address defaults in console from `host` to `hostname`

### CODE REVIEW

File: sdk/js/src/service/gateways.js [hunk 1]
```diff
@@ -74,11 +75,21 @@ class Gateways {
...
   // Create
 
-  async create (userId = this._defaultUserId, gateway) {
+  async create (userId = this._defaultUserId, gateway, { setDefaults = true }) {
+    let gtw = gateway
+    const Url = URL ? URL : window.URL
+
+    if (setDefaults) {
+      gtw = {
+        gateway_server_address: this._stackConfig.gs.enabled ? new Url(this._stackConfig.gs).hostname : '',
```
Comment (kschiffer) [inline sdk/js/src/service/gateways.js]:
```suggestion
        gateway_server_address: this._stackConfig.gs.enabled ? new Url(this._stackConfig.gs).hostname : '',
```

File: sdk/js/src/service/gateways.js [hunk 2]
```diff
@@ -74,11 +75,21 @@ class Gateways {
 
   // Create
 
-  async create (userId = this._defaultUserId, gateway) {
+  async create (userId = this._defaultUserId, gateway, { setDefaults = true }) {
+    let gtw = gateway
+    const Url = URL ? URL : window.URL
+
+    if (setDefaults) {
```
Comment (kschiffer) [inline sdk/js/src/service/gateways.js]: We decided to not use the any set default flag anymore in the SDK:
https://github.com/TheThingsNetwork/lorawan-stack/issues/683
The server address should rather be passed explicitly to avoid *magic*.

(It also still needs to be removed from the device service)

**Comment by htdvisser** [comment]
@kschiffer I fixed that, so I guess this PR now also closes #683.

Here's the updated device view:

Screenshot 2019-07-26 at 10 20 41

Maybe the field descriptions aren't really necessary, what do you think?

**Comment by htdvisser** [comment]
Making this a release blocker, since the current console creates end devices and gateways with different defaults than the CLI. We want that to be the same to avoid having to fix inconsistent databases.

File: pkg/webui/console/components/device-data-form/messages.js [hunk 3]
```diff
@@ -22,6 +22,9 @@ export default defineMessages({
   deviceIdDescription: 'Unique Identifier of your device; this cannot be changed afterwards',
   deviceNameDescription: 'Human friendly name of your device for display purposes',
   deviceDescDescription: 'Optional device description; can also be used to save notes about the device',
+  nsServerAddressDescription: 'The address of the device\'s Network Server',
```
Comment (kschiffer) [inline pkg/webui/console/components/device-data-form/messages.js]: I'm afraid this wording might be misleading, should we rather speak of `host` / `host name`? `address` sounds more like FQDN. (Same issue for gateway creation and linking form)

File: sdk/js/src/service/gateways.js [hunk 4]
```diff
@@ -79,7 +80,6 @@ class Gateways {
       routeParams: { 'collaborator.user_ids.user_id': userId },
     },
     { gateway })
-
```
Comment (kschiffer) [inline sdk/js/src/service/gateways.js]: Unnecessary change.

File: pkg/webui/console/views/gateway-add/index.js [hunk 5]
```diff
@@ -91,19 +91,12 @@ export default class GatewayAdd extends React.Component {
...
-      gsServerAddress = new URL(gs.base_url).host
-    }
-
     const initialValues = {
       ids: {
         gateway_id: undefined,
       },
       enforce_duty_cycle: true,
-      gateway_server_address: gsServerAddress,
+      gateway_server_address: env.config.gs.enabled ? new URL(env.config.gs.base_url).hostname : '',
```
Comment (kschiffer) [inline pkg/webui/console/views/gateway-add/index.js]:
```suggestion
      gateway_server_address: config.gs.enabled
        ? new URL(config.gs.base_url).hostname 
        : '',
```
See comment above.

File: sdk/js/src/service/devices/index.js [hunk 6]
```diff
@@ -252,18 +252,8 @@ class Devices {
     return this._responseTransform(response)
   }
 
-  async create (applicationId, device, { abp = false, setDefaults = true, withRootKeys = false } = {}) {
+  async create (applicationId, device, { abp = false, withRootKeys = false } = {}) {
     let dev = device
```
Comment (bafonins) [inline sdk/js/src/service/devices/index.js]:
```suggestion
    const dev = device
```

File: sdk/js/src/service/gateways.js [hunk 7]
```diff
@@ -12,6 +12,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+import { URL } from 'url'
```
Comment (bafonins) [inline sdk/js/src/service/gateways.js]: not needed

File: pkg/webui/console/views/device-add/index.js [hunk 8]
```diff
@@ -87,6 +89,13 @@ export default class DeviceAdd extends Component {
 
   render () {
     const { error } = this.state
+    const { env } = this.props
+
+    const initialValues = {
```
Comment (kschiffer) [inline pkg/webui/console/views/device-add/index.js]: Please refactor like this
```js
    const { env: { config }} = this.props

    const initialValues = {
      network_server_address: config.ns.enabled
        ? new URL(config.ns.base_url).hostname 
        : '',
        
        …
    }
```

File: pkg/webui/console/components/device-data-form/index.js [hunk 9]
```diff
@@ -179,6 +179,13 @@ class DeviceDataForm extends Component {
     const { update } = this.props
     return (
       <React.Fragment>
+        <Form.Field
+          title={sharedMessages.joinServerAddress}
+          description={m.jsServerAddressDescription}
+          placeholder={sharedMessages.addressPlaceholder}
```
Comment (kschiffer) [inline pkg/webui/console/components/device-data-form/index.js]: Please update the placeholder as well, it's still saying `host:port`

File: sdk/js/src/service/devices/index.js [hunk 10]
```diff
@@ -252,18 +252,8 @@ class Devices {
     return this._responseTransform(response)
   }
 
-  async create (applicationId, device, { abp = false, setDefaults = true, withRootKeys = false } = {}) {
```
Comment (kschiffer) [inline sdk/js/src/service/devices/index.js]: Please remove the `URL` import as well.

File: pkg/webui/console/components/device-data-form/validation-schema.js [hunk 11]
```diff
@@ -73,6 +73,12 @@ export default Yup.object().shape({
   lorawan_phy_version: Yup.string().required(sharedMessages.validateRequired),
   frequency_plan_id: Yup.string().required(sharedMessages.validateRequired),
   supports_class_c: Yup.boolean(),
+  network_server_address: Yup.string()
+    .matches(addressRegexp, sharedMessages.validateAddressFormat),
```
Comment (kschiffer) [inline pkg/webui/console/components/device-data-form/validation-schema.js]: The validation message also needs to be adapted then, it is still saying to use `host:port`

File: pkg/webui/console/components/device-data-form/index.js [hunk 12]
```diff
@@ -348,6 +354,18 @@ class DeviceDataForm extends Component {
           name="supports_class_c"
           component={Checkbox}
         />
+        <Form.Field
```
Comment (kschiffer) [inline pkg/webui/console/components/device-data-form/index.js]: Do you think it makes sense to put these under another sub-heading like `Connections` or is it good as part of `General Settings`?

File: pkg/webui/console/components/device-data-form/messages.js (hunk 3 above)
Comment (htdvisser) [reply pkg/webui/console/components/device-data-form/messages.js]: This was already removed as per @bafonins' suggestion

File: sdk/js/src/service/devices/index.js (hunk 10 above)
Comment (htdvisser) [reply sdk/js/src/service/devices/index.js]: Already done after @bafonins' review

File: pkg/webui/console/components/device-data-form/index.js (hunk 12 above)
Comment (htdvisser) [reply pkg/webui/console/components/device-data-form/index.js]: Not really. IMO everything stored on the Identity Server is "General".

Also, it's not actually connections (that's only for the GS)

**Comment by kschiffer** [comment]
> @kschiffer I fixed that, so I guess this PR now also closes #683.

Indeed.

**Comment by kschiffer** [comment]
> Maybe the field descriptions aren't really necessary, what do you think?

Agreed. At least the current ones are not really helpful. If you can think of something that comparably concise but increases clarity, go for it. Otherwise, I think we should remove this.
